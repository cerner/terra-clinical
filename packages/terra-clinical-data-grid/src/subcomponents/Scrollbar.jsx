import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import { DraggableCore } from 'react-draggable';

import styles from './Scrollbar.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * Function called once upon initial scroll of the scrollbar for a given movement event. Parameters: `onMoveStart()`
   */
  onMoveStart: PropTypes.func,
  /**
   * Function called once upon the end of scrolling of the scrollbar for a given movement event. Parameters: `onMoveEnd()`
   */
  onMoveEnd: PropTypes.func,
  /**
   * Function called upon movement of the scrollbar. Parameters: `onMove()`
   */
  onMove: PropTypes.func,
  /**
   * Function called with a reference to the Scrollbar's outer container element. Parameters: `refCallback(ref)`
   */
  refCallback: PropTypes.func,
  /**
   * Function called with a reference to the Scrollbar's scrollbar handle. Parameters: `scrollbarRefCallback(scrollbarRef)`
   */
  scrollbarRefCallback: PropTypes.func,
};

class Scrollbar extends React.Component {
  constructor(props) {
    super(props);

    this.handleDragStart = this.handleDragStart.bind(this);
    this.handleDragStop = this.handleDragStop.bind(this);

    this.scrollPosition = 0;
  }

  handleDragStart(event, data) {
    const { onMoveStart } = this.props;
    const handleNode = data.node;

    handleNode.parentNode.classList.add(cx('dragging'));

    if (onMoveStart) {
      onMoveStart(event, data);
    }
  }

  handleDragStop(event, data) {
    const { onMoveEnd } = this.props;
    const handleNode = data.node;

    handleNode.parentNode.classList.remove(cx('dragging'));

    if (onMoveEnd) {
      onMoveEnd(event, data);
    }
  }

  render() {
    const { onMove, refCallback, scrollbarRefCallback } = this.props;

    return (
      <div
        className={cx('container')}
        ref={refCallback}
      >
        <DraggableCore
          onStart={this.handleDragStart}
          onStop={this.handleDragStop}
          onDrag={onMove}
        >
          <div
            className={cx('scrollbar')}
            ref={scrollbarRefCallback}
          >
            <div
              className={cx('scrollbar-handle')}
            />
          </div>
        </DraggableCore>
      </div>
    );
  }
}

Scrollbar.propTypes = propTypes;

export default Scrollbar;
