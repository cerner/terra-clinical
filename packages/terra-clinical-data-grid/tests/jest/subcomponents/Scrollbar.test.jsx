import React from 'react';
import Scrollbar from '../../../src/subcomponents/Scrollbar';

describe('Scrollbar Snapshot Tests', () => {
  it('should render a Scrollbar with minimal props', () => {
    const scrollbar = shallow((
      <Scrollbar />
    ));

    expect(scrollbar).toMatchSnapshot();
  });

  it('should render a Scrollbar with all provided props', () => {
    const scrollbar = shallow((
      <Scrollbar
        onMoveStart={() => {}}
        onMoveEnd={() => {}}
        onMove={() => {}}
        refCallback={() => {}}
        scrollbarRefCallback={() => {}}
      />
    ));

    expect(scrollbar).toMatchSnapshot();
  });
});
