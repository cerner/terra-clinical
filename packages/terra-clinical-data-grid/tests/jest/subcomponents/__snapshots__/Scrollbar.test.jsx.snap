// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scrollbar Snapshot Tests should render a Scrollbar with all provided props 1`] = `
<div
  className="container"
>
  <DraggableCore
    allowAnyClick={false}
    cancel={null}
    disabled={false}
    enableUserSelectHack={true}
    grid={null}
    handle={null}
    offsetParent={null}
    onDrag={[Function]}
    onMouseDown={[Function]}
    onStart={[Function]}
    onStop={[Function]}
    transform={null}
  >
    <div
      className="scrollbar"
    >
      <div
        className="scrollbar-handle"
      />
    </div>
  </DraggableCore>
</div>
`;

exports[`Scrollbar Snapshot Tests should render a Scrollbar with minimal props 1`] = `
<div
  className="container"
>
  <DraggableCore
    allowAnyClick={false}
    cancel={null}
    disabled={false}
    enableUserSelectHack={true}
    grid={null}
    handle={null}
    offsetParent={null}
    onDrag={[Function]}
    onMouseDown={[Function]}
    onStart={[Function]}
    onStop={[Function]}
    transform={null}
  >
    <div
      className="scrollbar"
    >
      <div
        className="scrollbar-handle"
      />
    </div>
  </DraggableCore>
</div>
`;
